From e57fe8fc9b2b01b407659bf85b30b04676e5c7a9 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 04 Nov 2024 09:51:28 +0200
Subject: [PATCH] [FSW-2071] fs1: add auto image loading params

Images can be loaded automatically from flash by using data from the
image manifest. To enable that some additional parameters are added
to the manifest, such as memory address & size, cpu details, and
image flags.
Since Preboot is the only image currently loaded in an automated manner,
the new parameters are currently initialized only for Preboot

Change-Id: I228434317c7c1950a8fb39401c3ad495463088ca
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/board/habanalabs/fs1_arc/post-image-spi.sh b/board/habanalabs/fs1_arc/post-image-spi.sh
index d8b63bc..40077f0 100755
--- a/board/habanalabs/fs1_arc/post-image-spi.sh
+++ b/board/habanalabs/fs1_arc/post-image-spi.sh
@@ -29,6 +29,8 @@
 PREBOOT_SEC_OFFSET=$((PREBOOT_PRI_OFFSET + SECONDARY_MANIFEST_OFFSET))
 PREBOOT_TEXT_BASE=0x300007FF8A08000
 PREBOOT_START_ADDR=0x300007FF8A08000
+PREBOOT_CPU_BASE_ADDR=0x79CE0000	# mmIO_0_CFG_HALF_0_PSOC_ARC_MNG_ARC_BASE
+PREBOOT_IMAGE_FLAGS=0x7E		# load and exec for all rst types
 
 FAKE_FUSE_PRI_OFFSET=0x140000
 FAKE_FUSE_SEC_OFFSET=$((FAKE_FUSE_PRI_OFFSET + SECONDARY_MANIFEST_OFFSET))
@@ -419,6 +421,12 @@
 	local img_dma_alt_h_1=${img_loadh}
 	local img_dma_alt_l_1=${img_loadl}
 	local img_name_2=$img_name
+	local img_flags=0
+	local img_cpu_base=0
+	local img_mem_base_l=0
+	local img_mem_base_h=0
+	local img_mem_size=0
+
 	if [ $img_name == "PPBOOT" ]
 	then
 		img_name_2="PPBOOT_PCIE_PHY_PADDED"
@@ -433,6 +441,8 @@
 	then
 		img_base_pri=${spi_reg_offsets[PREBOOT_IO_DIE_PRIMARY]}
 		img_base_sec=${spi_reg_offsets[PREBOOT_IO_DIE_SECONDARY]}
+		img_cpu_base=${PREBOOT_CPU_BASE_ADDR}
+		img_flags=${PREBOOT_IMAGE_FLAGS}
 	fi
 
 	if [ $img_name == "FAKE_FUSE" ]
@@ -505,6 +515,13 @@
 
 	# Prepare component version
 	print_image_version $img_name
+
+	# Prepare autoload parameters
+	printf "%08x" $img_cpu_base | rev_bytes_stdin
+	printf "%08x" $img_mem_base_h | rev_bytes_stdin
+	printf "%08x" $img_mem_base_l | rev_bytes_stdin
+	printf "%08x" $img_mem_size | rev_bytes_stdin
+	printf "%08x" $img_flags | rev_bytes_stdin
 }
 
 function prep_image_ptrs_common_v1 {
